prediction_market_agent/agents/safe_guard_agent/app.py

```diff
@@ -47,31 +53,89 @@
     "Your safe address (or keep the default one provided)",
     value="gno:0x1275EE969cd187Fd7B9c6Aa8339d41985eFd1EF1",
 )
-do_execute = st.checkbox("Execute transaction if validated", value=False)
-do_reject = st.checkbox("Reject transaction if not validated", value=False)
-do_message = st.checkbox("Send a message about the outcome", value=False)
 
 if not safe_address:
     st.stop()
 
 # Get rid of potential token id
 *_, safe_address = safe_address.split(":")
 safe_address_checksum = Web3.to_checksum_address(safe_address)
+is_owner = check_if_owner(safe_address_checksum, keys.bet_from_address)
+
+do_execute = st.checkbox(
+    "Execute transaction if validated (possible only for Safe's owners)",
+    value=False,
+    disabled=not is_owner,
+)
+do_reject = st.checkbox(
+    "Reject transaction if not validated (possible only for Safe's owners)",
+    value=False,
+    disabled=not is_owner,
+)
+do_message = st.checkbox(
+    "Send a message about the outcome (possible only for Safe's owners)",
+    value=False,
+    disabled=not is_owner,
+)
+
+
+@st.cache_data(ttl=60)
+def get_safe_queued_transactions_cached(
+    safe_address_checksum: ChecksumAddress,
+) -> list[safe_api_utils.Transaction]:
+    return safe_api_utils.get_safe_queued_transactions(safe_address_checksum)
+
 
-with st.spinner("Loading queued transactions..."):
-    queued_transactions = safe_api_utils.get_safe_queued_transactions(
-        safe_address_checksum
+@st.cache_data(ttl=60)
+def get_safe_history_cached(
+    safe_address_checksum: ChecksumAddress,
+) -> list[safe_api_utils.Transaction]:
+    return safe_api_utils.get_safe_history(safe_address_checksum)
+
+
+queued_transactions = get_safe_queued_transactions_cached(safe_address_checksum)
+historical_transactions = get_safe_history_cached(safe_address_checksum)
+
+queue_col, hist_col = st.columns(2)
+with queue_col:
+    queue_transaction_id = st.selectbox(
+        "Choose one of the queued transactions",
+        [tx.id for tx in queued_transactions],
+        index=None,
+    )
+with hist_col:
+    hist_transaction_id = st.selectbox(
+        "Or historical transaction for simulation (available if no queue transaction is selected)",
+        [tx.id for tx in historical_transactions],
+        index=None,
+        disabled=queue_transaction_id is not None,
     )
 
-transaction_id = st.selectbox(
-    "Choose one of the queued transactions",
-    [tx.id for tx in queued_transactions],
+transaction_id = queue_transaction_id or hist_transaction_id
+if transaction_id is None:
+    st.stop()
+
+transaction = check_not_none(
+    next(
+        (
+            tx
+            for tx in queued_transactions + historical_transactions
+            if tx.id == transaction_id
+        ),
+        None,
+    )
```

Comment:
Sorry, why was the previous implementation not good enough? I didn't follow this change.

Reply:
Previously, `transaction_id` was passed to the function, but because `queued` and `historical` transactions are two separate endpoints, it's not enough to pass only the id.

Also if we have the whole object here, why not just pass it instead of pass the ID + re-query the endpoint.